### ISSUE TITLE
Does GanEstimator support multi GPU ?

### ISSUE THREAD

**Issue by paigeQ** [tensorflow/tensorflow#23990]
Does GanEstimator support multi gpu?
How could I using multi GPU by ganEstimator?

Thanks.

**Comment by nwestlake** [comment]
I found the only issue was the need to change check_for_unused_update_ops to False in [tensorflow/contrib/gan/python/train.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/gan/python/train.py)

Of course the code to that, without changing the source, is very long:

    # Custom estimator classes
    class GanEstimatorTowerOpSafe(tf.contrib.gan.estimator.GANEstimator):
        """The only change is the use of new_get_estimator_spec"""
        def __init__(self,
                     model_dir=None,
                     generator_fn=None,
                     discriminator_fn=None,
                     generator_loss_fn=None,
                     discriminator_loss_fn=None,
                     generator_optimizer=None,
                     discriminator_optimizer=None,
                     get_hooks_fn=None,
                     get_eval_metric_ops_fn=None,
                     add_summaries=None,
                     use_loss_summaries=True,
                     config=None,
                     warm_start_from=None,
                     is_chief=True):
            if not callable(generator_loss_fn):
                raise ValueError('generator_loss_fn must be callable.')
            if not callable(discriminator_loss_fn):
                raise ValueError('discriminator_loss_fn must be callable.')
            if use_loss_summaries not in [True, False, None]:
                raise ValueError('use_loss_summaries must be True, False or None.')
            if get_hooks_fn is not None and not callable(get_hooks_fn):
                raise TypeError('get_hooks_fn must be callable.')

            from tensorflow.contrib.gan.python.estimator.python.gan_estimator_impl\
                import _get_gan_model

            def _model_fn(features, labels, mode):
                """GANEstimator model function."""
                from tensorflow.python.estimator import model_fn as model_fn_lib
                if mode not in [model_fn_lib.ModeKeys.TRAIN,
                                model_fn_lib.ModeKeys.EVAL,
                                model_fn_lib.ModeKeys.PREDICT]:
                    raise ValueError('Mode not recognized: %s' % mode)
                real_data = labels    # rename inputs for clarity
                generator_inputs = features    # rename inputs for clarity

                gan_model = _get_gan_model(
                    mode, generator_fn, discriminator_fn, real_data,
                    generator_inputs, add_summaries)

                return new_get_estimator_spec(
                    mode, gan_model, generator_loss_fn, discriminator_loss_fn,
                    get_eval_metric_ops_fn, generator_optimizer,
                    discriminator_optimizer,
                    get_hooks_fn, use_loss_summaries, is_chief)

            super(tf.contrib.gan.estimator.GANEstimator, self).__init__(
                model_fn=_model_fn, model_dir=model_dir, config=config,
                warm_start_from=warm_start_from)


    def new_get_estimator_spec(
        mode, gan_model, generator_loss_fn, discriminator_loss_fn,
            get_eval_metric_ops_fn, generator_optimizer, discriminator_optimizer,
            get_hooks_fn=None, use_loss_summaries=True, is_chief=True):
        # The only difference and hence fix is check_for_unused_update_ops=False

        from tensorflow.python.estimator import model_fn as model_fn_lib
        from tensorflow.contrib.gan.python import namedtuples as tfgan_tuples
        from tensorflow.contrib.gan.python import train as tfgan_train
        from tensorflow.contrib.gan.python.estimator.python.gan_estimator_impl\
            import _get_eval_estimator_spec, _get_train_estimator_spec

        if mode == model_fn_lib.ModeKeys.PREDICT:
            estimator_spec = model_fn_lib.EstimatorSpec(
                mode=mode, predictions=gan_model.generated_data)
        else:
            gan_loss = tfgan_tuples.GANLoss(
                generator_loss=generator_loss_fn(
                    gan_model, add_summaries=use_loss_summaries),
                discriminator_loss=discriminator_loss_fn(
                    gan_model, add_summaries=use_loss_summaries))
            if mode == model_fn_lib.ModeKeys.EVAL:
                estimator_spec = _get_eval_estimator_spec(
                    gan_model, gan_loss, get_eval_metric_ops_fn)
            else:  # model_fn_lib.ModeKeys.TRAIN:
                gopt = (generator_optimizer() if callable(generator_optimizer) else
                        generator_optimizer)
                dopt = (discriminator_optimizer() if callable(
                    discriminator_optimizer) else discriminator_optimizer)
                get_hooks_fn = (get_hooks_fn or
                                tfgan_train.get_sequential_train_hooks())

                def new_train_op_fn(model, loss, generator_optimizer,
                                    discriminator_optimizer, is_chief=True):
                    return tfgan_train.gan_train_ops(
                        model, loss, generator_optimizer, discriminator_optimizer,
                        check_for_unused_update_ops=False, is_chief=is_chief)

                estimator_spec = _get_train_estimator_spec(
                    gan_model, gan_loss, gopt, dopt, get_hooks_fn,
                    train_op_fn=new_train_op_fn, is_chief=is_chief)

        return estimator_spec

**Comment by tridemax** [comment]
@nwestlake will it affect the training as other's tower update ops (like batch norm) will not get called? I see explicit filtering of update ops like 'tower_1/Generator/...'

**Comment by blackPython** [comment]
I don't think it supported. Because for training generator and discriminator at different ratios (as in update generator and discriminator at different rate) they have included the main training steps, i.e. update of discriminator and generator as training hooks. But in case of estimator with multiGPU training, the trainning hooks from only the first replica (Tower) are run  ( Can be seen in this [line](https://github.com/tensorflow/estimator/blob/r1.13/tensorflow_estimator/python/estimator/estimator.py#L1269)). Hence even if u are using the multiple GPUs the training steps will be taken only once.

**Comment by rmothukuru** [comment]
@paigeQ,
Can you please let us know if this issue is still relevant because there is no documentation corresponding to GAN Estimator in Tensorflow Documentation. Also, please refer the documentation of [Distribution Strategy for Estimator](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_estimator), just in case if it helps. Thanks!
